704 raspberry usb

mike2390 opened the pull request:
This PR includes:
- Initial version of the USB driver for RPi
- Moved mailbox and memory barriers outside of the video driver, because they
  are used also for other purposes

Note: currently the USB is not working. The controller does not send
interrupts after setup.

anton-bondarev commented:
LGTM

abusalimov commented on src/drivers/mailbox/bcm2835_mailbox.c:
```diff
@@ -0,0 +1,73 @@
+{
+	data_mem_barrier();
+	REG_STORE(addr, val);
+	data_mem_barrier();
```
Is the second barrier really needed? [BCM2835 Datasheet](http://www.farnell.com/datasheets/1521578.pdf) suggests that barriers should be placed as follows:

> • A memory write barrier before the first write to a peripheral.
> • A memory read barrier after the last read of a peripheral.

mike2390 replied:
[Here](http://infocenter.arm.com/help/topic/com.arm.doc.dai0321a/DAI0321A_programming_guide_memory_barriers_for_m_profile.pdf) at chapter 3.1 they say:

Data Memory Barrier
 Ensures that all explicit data memory transfers before the DMB are completed before any subsequent data memory transfers after the DMB starts.

So we can follow the guidelines from the BCM2835 Datasheeet and ensure all data memory transfers before write are completed before our write to the mailbox. For the read we ensure that all data memory accesses plus the read from the mailbox are completed before the DMB.

abusalimov replied:
I'm not quite sure I understand you. Following the BCM2835 guide means leaving only a barrier _before_ `REG_STORE` and a barrier _after_ `REG_LOAD`, doesn't it?

mike2390 replied:
Exactly, in this way you ensure what I described before :). I will modify that as well.

abusalimov replied:
Ah, yeah, that's actually what I meant originally. :)
Thanks! :+1:

abusalimov commented on src/arch/arm/armlib/mem_barriers.S:
```diff
@@ -0,0 +1,31 @@
+.global data_mem_barrier
+data_mem_barrier:
+    mov r12, #0
+    mcr p15, 0, r12, c7, c10, 5
```
I would rather implement it as an inline function, in similar way like we do for `ipl_xxx` operations:
```c
static inline void data_memory_barrier(void) {
	__asm__ __volatile__("mcr p15, 0, %0, c7, c10, 5"
			: : "r" (0) : "memory");
}
```

mike2390 replied:
I wrote it directly in assembly because the body of the function is all written in assembly in the end. If you think it is better to have it inline, also for consistency with already existing routines, I can change it.

abusalimov replied:
Mostly for consistency, yes, thank you.

abusalimov commented:
Sorry for not reviewing your follow up changes for that long, I didn't notice you've already made them. I wish GitHub could notify about new commits within a PR somehow... Anyway, don't hesitate to drop a ping next time you see no one responds. ;-)

mike2390 commented:
No worries, I thought you were busy :). Could you check now if everything is fine also with the deleted files raspi_mailbox?

Thanks

abusalimov commented:
That's not quite what I meant. I pushed that into my fork, please check it out:
```
git branch backup-704-raspberry-usb
git reset --hard master
git pull https://github.com/abusalimov/embox.git mike2390-704-raspberry-usb
```

mike2390 commented:
Do you want to obtain something like the following?

```
renamed src/drivers/video/raspi_mailbox.h -> src/drivers/mailbox/bcm2835_mailbox.h
renamed src/drivers/video/raspi_mailbox.c -> src/drivers/mailbox/bcm2835_mailbox.c
```

abusalimov commented:
Yes, basically. It is also needed for e.g. the "Show origin of this line" gitk command.

abusalimov commented:
> That's not quite what I meant. I pushed that into my fork, please check it out:
> ```
> git branch backup-704-raspberry-usb
> git reset --hard master
> git pull https://github.com/abusalimov/embox.git mike2390-704-raspberry-usb
> ```

P.S. If you find it OK, just `push origin 704-raspberry-usb --force`.

mike2390 commented:
I cannot manage to force git to do the renaming. Is it OK if I go back to ```raspi: Separated mailbox from raspi video driver``` and do ```git rm``` on those files?

mike2390 commented:
Never mind I got it ;-)

abusalimov commented:
It may depend on git-diff sensitivity settings. Yes, that would be enough, in fact, that's what I did, plus I also moved some irrelevant changes there too.
